### PR TITLE
[web][SIMS #487] Ministry Views Student Application Details - Ministry user viewing application form

### PR DESCRIPTION
**Pull request by ann-aot**
- fixed Ministry user viewing read-only viewing application form
- did some refactoring in sharing the application summary component
- actions are triggered by events from the shared component
- 'offeringIntensity` was not working properly in student forms
-**_Screenshots_**
image

### CODE REVIEW

File: sources/packages/api/src/services/education-program/education-program.service.ts [hunk 1]
```diff
@@ -431,17 +428,21 @@ export class EducationProgramService extends RecordDataModelService<EducationPro
...
-      })
       .andWhere(`exists(${offeringExistsQuery.getQuery()})`)
       .select("programs.id")
       .addSelect("programs.name")
       .setParameters({ locationId, programYearId })
-      .orderBy("programs.name")
-      .getMany();
+      .orderBy("programs.name");
+
+    if (!includeInActivePY) {
```
Comment (andrewsignori-aot) [inline sources/packages/api/src/services/education-program/education-program.service.ts:440]: Why a variable called includeInActivePY is defining the ApprovalStatus?

File: sources/packages/web/src/components/aest/StudentApplications.vue [hunk 2]
```diff
@@ -125,53 +124,49 @@ import {
...
+      required: false,
+      default: false,
+    },
+    reloadData: {
+      type: Boolean,
+      default: false,
     },
   },
-  setup(props: any) {
+  setup(_props: any) {
```
Comment (andrewsignori-aot) [inline sources/packages/web/src/components/aest/StudentApplications.vue:155]: Please name _props_ accordingly.

File: sources/packages/web/src/views/aest/student/Applications.vue [hunk 3]
```diff
@@ -17,5 +22,19 @@ export default {
       required: true,
     },
   },
+  setup(_props: any) {
```
Comment (andrewsignori-aot) [inline sources/packages/web/src/views/aest/student/Applications.vue:25]: Please name props accordingly.

File: sources/packages/web/src/views/aest/student/Applications.vue [hunk 4]
```diff
@@ -17,5 +22,19 @@ export default {
...
+        name: AESTRoutesConst.APPLICATION_DETAILS,
+        params: {
+          applicationId: id,
+          studentId: _props.studentId,
+        },
+      });
+    };
+    return { goToApplication };
+  },
 };
```
Comment (henryoforeh-aot) [inline sources/packages/web/src/views/aest/student/Applications.vue:39]: router.push is a promise. You may need to use async/await in this function.

File: sources/packages/web/src/views/student/StudentApplicationSummary.vue [hunk 5]
```diff
@@ -38,14 +57,86 @@ export default {
...
+    const getProgramYear = async (applicationId: number) => {
+      programYear.value = await ApplicationService.shared.getProgramYearOfApplication(
+        applicationId,
+      );
+    };
+
+    const editApplication = async (applicationId: number) => {
+      try {
+        await getProgramYear(applicationId);
+        router.push({
```
Comment (henryoforeh-aot) [inline sources/packages/web/src/views/student/StudentApplicationSummary.vue:93]: please use await before router.push.

File: sources/packages/web/src/views/student/StudentApplicationSummary.vue [hunk 6]
```diff
@@ -38,14 +57,86 @@ export default {
...
+    };
+
+    const goToApplication = (id: number) => {
+      return router.push({
+        name: StudentRoutesConst.STUDENT_APPLICATION_DETAILS,
+        params: {
+          id: id,
+        },
+      });
+    };
```
Comment (henryoforeh-aot) [inline sources/packages/web/src/views/student/StudentApplicationSummary.vue:136]: router.push is a promise. Please refer to my previous comment on using async/await.

File: sources/packages/web/src/components/aest/StudentApplications.vue [hunk 7]
```diff
@@ -26,7 +28,7 @@
         <v-btn
           v-if="clientType === ClientIdType.Student"
           plain
-          @click="goToApplication(slotProps.data.id)"
+          @click="$emit('goToApplication', slotProps.data.id)"
```
Comment (dheepak-aot) [inline sources/packages/web/src/components/aest/StudentApplications.vue:31]: I see emit being used multiple times, We can create an emit function and call from all the click.

File: sources/packages/web/src/types/AppRoutes.ts [hunk 8]
```diff
@@ -54,7 +54,7 @@ export enum AppRoutes {
   Restrictions = "restrictions",
   Notes = "notes",
   Applications = "applications",
-  ApplicationDetail = "application-detail",
+  ApplicationDetail = "student/:studentId/application/:applicationId",
```
Comment (dheepak-aot) [inline sources/packages/web/src/types/AppRoutes.ts:57]: 👍

File: sources/packages/web/src/views/aest/student/Applications.vue (hunk 4 above)
Comment (dheepak-aot) [reply sources/packages/web/src/views/aest/student/Applications.vue:38]: As per Vue, if we are waiting for the result of navigation which can be (NavigationFailure | void | undefined ) and do something with the result then await is useful. IMO otherwise it should be fine without it. @ann-aot @andrewsignori-aot thoughts?

File: sources/packages/web/src/components/aest/StudentApplications.vue (hunk 7 above)
Comment (ann-aot) [reply sources/packages/web/src/components/aest/StudentApplications.vue:31]: `$emit` is the vue3 in-build function 
image
.it avoids additions code in setup()

File: sources/packages/web/src/views/aest/student/Applications.vue (hunk 4 above)
Comment (ann-aot) [reply sources/packages/web/src/views/aest/student/Applications.vue:38]: I believe, if are doing anything after the router, then we need to wait, else I think there is no need. correct me if I am wrong @dheepak-aot @henryoforeh-aot  @andrewsignori-aot 
image
https://router.vuejs.org/guide/advanced/navigation-failures.html

File: sources/packages/web/src/views/student/StudentApplicationSummary.vue [hunk 9]
```diff
@@ -38,14 +57,86 @@ export default {
...
+    ConfirmEditApplication,
+    CancelApplication,
   },
   setup() {
     const hasRestriction = ref(false);
     const restrictionMessage = ref("");
+    const programYear = ref({} as ProgramYearOfApplicationDto);
+    const router = useRouter();
+    const toast = useToastMessage();
+    const TOAST_ERROR_DISPLAY_TIME = 15000;
```
Comment (dheepak-aot) [inline sources/packages/web/src/views/student/StudentApplicationSummary.vue:69]: can create a global constant for 15000 and we can use it from there.

File: sources/packages/web/src/views/student/StudentApplicationSummary.vue [hunk 10]
```diff
@@ -38,14 +57,86 @@ export default {
...
+          name: StudentRoutesConst.DYNAMIC_FINANCIAL_APP_FORM,
+          params: {
+            selectedForm: programYear.value.formName,
+            programYearId: programYear.value.programYearId,
+            id: applicationId,
+          },
+        });
+      } catch (error) {
+        toast.error(
+          "Program Year not active",
```
Comment (dheepak-aot) [inline sources/packages/web/src/views/student/StudentApplicationSummary.vue:103]: If there is an error on getProgramYear() API call, can't there be any other reason except program year not being active.

File: sources/packages/web/src/views/student/financial-aid-application/FullTimeApplication.vue [hunk 11]
```diff
@@ -373,16 +377,19 @@ export default {
...
-        );
+        if (selectedLocationId) {
+          // when isReadOnly.value is true, then consider
+          // both active and inactive program year.
+          await formioDataLoader.loadProgramsForLocation(
+            form,
+            +selectedLocationId,
+            PROGRAMS_DROPDOWN_KEY,
+            props.programYearId,
+            isReadOnly.value,
```
Comment (dheepak-aot) [inline sources/packages/web/src/views/student/financial-aid-application/FullTimeApplication.vue:390]: If we are going to do label for selected program, we can ignore this part right. Correct me if I am missing something.

File: sources/packages/web/src/components/aest/StudentApplications.vue (hunk 7 above)
Comment (dheepak-aot) [reply sources/packages/web/src/components/aest/StudentApplications.vue:31]: Got it. Thanks 👍

File: sources/packages/web/src/views/student/StudentApplicationSummary.vue (hunk 10 above)
Comment (ann-aot) [reply sources/packages/web/src/views/student/StudentApplicationSummary.vue:103]: I believe, this API is looking for an active program year while editing, 
image


these vue changes are part of refactoring, I move this function  from the shared component to the student's parent component calling the shared component

**Comment by dheepak-aot** [comment]
LGTM @ann-aot. 👍2 minor comments.

File: sources/packages/web/src/views/student/financial-aid-application/FullTimeApplication.vue (hunk 11 above)
Comment (ann-aot) [reply sources/packages/web/src/views/student/financial-aid-application/FullTimeApplication.vue:390]: `label` is not part of these PR. and this logic or codes are not added by me. It already, I only added some conditions and some different logic to resolve the program intensity refreshing bug in the application form, 
image

**Review by henryoforeh-aot — APPROVED** [review]

File: sources/packages/web/src/views/student/StudentApplicationSummary.vue (hunk 10 above)
Comment (dheepak-aot) [reply sources/packages/web/src/views/student/StudentApplicationSummary.vue:103]: As discussed we can note this down and add to technical debt and dev discussion point at our next scheduled sync up meeting.

**Review by dheepak-aot — APPROVED** [review]
LGTM @ann-aot.

File: sources/packages/web/src/views/student/StudentApplicationSummary.vue (hunk 10 above)
Comment (ann-aot) [reply sources/packages/web/src/views/student/StudentApplicationSummary.vue:103]: added this note in https://app.zenhub.com/workspaces/student-information-management-system-5fce9df5aa1b45000e937014/issues/bcgov/sims/695

File: sources/packages/web/src/constants/message-constants.ts [hunk 12]
```diff
@@ -9,3 +9,8 @@ export const PENDING_SIN_MESSAGE =
  */
 export const INVALID_SIN_MESSAGE =
   "StudentAid BC has found an issue with your account information. Please ensure your account information matches CRA and log back in. If you have further issues please contact us.";
+
+/**
+ * toast display time
+ */
+export const TOAST_ERROR_DISPLAY_TIME = 15000;
```
Comment (andrewsignori-aot) [inline sources/packages/web/src/constants/message-constants.ts:16]: Can we move it inside the useToastMessage.ts and maybe called it EXTENDED_MESSAGE_DISPLAY_TIME

File: sources/packages/web/src/views/student/StudentApplicationSummary.vue [hunk 13]
```diff
@@ -38,14 +58,85 @@ export default {
     StudentApplications,
     CheckValidSINBanner,
     HeaderNavigator,
+    ConfirmEditApplication,
+    CancelApplication,
   },
   setup() {
     const hasRestriction = ref(false);
     const restrictionMessage = ref("");
+    const programYear = ref({} as ProgramYearOfApplicationDto);
```
Comment (andrewsignori-aot) [inline sources/packages/web/src/views/student/StudentApplicationSummary.vue:67]: Program year does not need to be a reactive property in this scenario.

File: sources/packages/web/src/views/student/StudentApplicationSummary.vue [hunk 14]
```diff
@@ -38,14 +58,85 @@ export default {
...
+      showModal.value = true;
+      selectedApplicationId.value = id;
+    };
+
+    const showHideCancelApplication = () => {
+      showModal.value = !showModal.value;
+    };
+
+    const getProgramYear = async (applicationId: number) => {
+      programYear.value = await ApplicationService.shared.getProgramYearOfApplication(
```
Comment (andrewsignori-aot) [inline sources/packages/web/src/views/student/StudentApplicationSummary.vue:85]: The method can return the program year id and avoid creating a unnecessary local variable.
For this scenario, I don't think that we really need a method to do it, _getProgramYearOfApplication_ can be called directly inside the _editApplication_.
Going further, not part of this PR, I believe that we should remove the need of having the program year in order to call application.

**Review by andrewsignori-aot — COMMENTED** [review]
Looks good, please take a look at the comments 😉

File: sources/packages/web/src/views/student/StudentApplicationSummary.vue (hunk 14 above)
Comment (ann-aot) [reply sources/packages/web/src/views/student/StudentApplicationSummary.vue:85]: Andrew, `editApplication` is actually doing go to Edit Application page(maybe I will rename the function to avoid confusion) and from this API 
image
its getting formname and programyear id
image
. this variable is confusing, i will rename the variable

Comment (ann-aot) [reply sources/packages/web/src/views/student/StudentApplicationSummary.vue]: but I got the point here whenever there is only need to id, we can create an API only to return ID

**Review by andrewsignori-aot — APPROVED** [review]
Thanks for doing the adjustments. Looks good 👍
